File: ruby/day11/part2.rb
# frozen_string_literal: true

def parse_input(input)
  input.split.map(&:to_i)
end

def blink_recursive(value, remaining_blinks, cache)
  return 1 if remaining_blinks.zero?

  cache_key = [value, remaining_blinks]
  return cache[cache_key] if cache.key?(cache_key)

  result = if value.zero?
             blink_recursive(1, remaining_blinks - 1, cache)
           else
             value_str = value.to_s
             if value_str.length.even?
               mid = value_str.length / 2
               first_half = value_str[0...mid].to_i
               second_half = value_str[mid..].to_i
               blink_recursive(first_half, remaining_blinks - 1, cache) +
                 blink_recursive(second_half, remaining_blinks - 1, cache)
             else
               blink_recursive(2024 * value, remaining_blinks - 1, cache)
             end
           end

  cache[cache_key] = result
  result
end

def blink(stones, blinks)
  cache = {}
  stones.reduce(0) { |acc, x| acc + blink_recursive(x, blinks, cache) }
end

def part2(input, blinks)
  numbers = parse_input(input)
  blink(numbers, blinks)
end

if __FILE__ == $PROGRAM_NAME
  input = File.read('input.txt')
  puts part2(input, 75)
end

File: elixir/day11/lib/part2.ex
defmodule Part2 do
  def parse_input(input) do
    input
    |> String.split()
    |> Enum.map(&String.to_integer/1)
  end

  def blink(stones, blinks) do
    cache = :ets.new(:blink_cache, [:set])

    result =
      Enum.reduce(stones, 0, fn x, acc ->
        acc + blink_recursive(x, blinks, cache)
      end)

    :ets.delete(cache)
    result
  end

  defp blink_recursive(_, 0, _cache), do: 1

  defp blink_recursive(value, remaining_blinks, cache) do
    cache_key = {value, remaining_blinks}

    case :ets.lookup(cache, cache_key) do
      [{^cache_key, result}] ->
        result

      [] ->
        result = calculate_blink(value, remaining_blinks, cache)
        :ets.insert(cache, {cache_key, result})
        result
    end
  end

  defp calculate_blink(0, remaining_blinks, cache) do
    blink_recursive(1, remaining_blinks - 1, cache)
  end

  defp calculate_blink(value, remaining_blinks, cache) do
    value_str = Integer.to_string(value)

    if rem(String.length(value_str), 2) == 0 do
      mid = div(String.length(value_str), 2)
      {first_half, second_half} = String.split_at(value_str, mid)
      first_num = String.to_integer(first_half)
      second_num = String.to_integer(second_half)

      blink_recursive(first_num, remaining_blinks - 1, cache) +
        blink_recursive(second_num, remaining_blinks - 1, cache)
    else
      blink_recursive(2024 * value, remaining_blinks - 1, cache)
    end
  end

  def part2(input, blinks) do
    input
    |> parse_input()
    |> blink(blinks)
  end
end

File.read!("input.txt")
|> Part2.part2(75)
|> IO.inspect(label: "part2")
